### PR TITLE
Add support for representing higher-order relationship types

### PR DESCRIPTION
**Pull request by pinnouse**
This draft pull request currently adds support for the diseq relation, as well as fixing up some of the current code base.

Ongoing: to implement the other relations to be represented

### CODE REVIEW

File: common.rkt [hunk 1]
```diff
@@ -167,7 +167,7 @@
 (struct Ans (term constraint) #:prefab)
 
 (define (walk* tm st)
-  (let* ((sub (state-sub st)) (tm (walk tm sub)))
+  (let* ((sub (if (state? st) (state-sub st) st)) (tm (walk tm sub)))
```
Comment (gregr) [inline common.rkt:170]: `walk*` doesn't need a state, just a substitution.  Let's change the callers to always pass a substitution.

File: tools.rkt [hunk 2]
```diff
@@ -71,7 +71,8 @@
      (let ((t1 (prune/term t1)) (t2 (prune/term t2)))
        (match (unify t1 t2 st)
          (#f          #f)
-         (`(,st . #f) (pause st (== t1 t2))))))))
+         (`(,st . #f) (pause st (== t1 t2)))
```
Comment (gregr) [inline tools.rkt:75]: `unify` should never return a pair.  So we only need the new case you added.

File: examples.rkt [hunk 3]
```diff
@@ -17,72 +17,73 @@
...
-    ((== '() ab) (== c abc))
-    ((fresh (a b bc)
-       (== `(,a . ,b) ab)
-       (== `(,a . ,bc) abc)
-       (appendo b c bc)))))
+   ((== '() ab) (== c abc))
+   ((fresh (a b bc)
+           (== `(,a . ,b) ab)
+           (== `(,a . ,bc) abc)
+           (appendo b c bc)))))
```
Comment (gregr) [inline examples.rkt:24]: `fresh` is a variable-binding form (similar to `let` or `lambda`), so its body should not be indented so much.  Can you adjust your editor's config to have it use the original indentation?

File: examples.rkt [hunk 4]
```diff
@@ -17,72 +17,73 @@
...
-  (run* (q) (appendo '(1 2 3) '(4 5) q))
-  (run* (p q) (appendo p q '(1 2 3 4 5)))
-  ;; We can change our search strategy by choosing a different way to step.
-  ;; `step` gives us a typical interleaving search.
-  (run*/step step (p q) (appendo p q '(1 2 3 4 5))))
+ (run* (q) (appendo '(1 2 3) '(4 5) q))
+ (run* (p q) (appendo p q '(1 2 3 4 5)))
+ ;; We can change our search strategy by choosing a different way to step.
+ ;; `step` gives us a typical interleaving search.
+ (run*/step step (p q) (appendo p q '(1 2 3 4 5)))
```
Comment (gregr) [inline examples.rkt:31]: Is it possible to adjust your editor to use 2-space indents in this situation?

File: examples.rkt [hunk 5]
```diff
@@ -17,72 +17,73 @@
...
+ (run*/step step (q r) (=/= q '()) (=/= q '(1)) (=/= q '(1 2)) (=/= q r) (appendo q r '(1 2 3 1 2 3))))
 
 
 (define-relation (reverseo ys sy)
   (conde
-    ((== '() ys) (== '() sy))
-    ((fresh (first rest prefix)
```
Comment (gregr) [inline examples.rkt:37]: 2-space indents for `conde` as well.

File: tools.rkt [hunk 6]
```diff
@@ -72,7 +72,13 @@
        (match (unify t1 t2 st)
          (#f          #f)
          (`(,st . #f) (pause st (== t1 t2)))
-         (st          (pause st (== t1 t2))))))))
+         (st          (pause st (== t1 t2))))))
+    ((=/= t1 t2)
+     (let ((t1 (prune/term t1)) (t2 (prune/term t2)))
+       (match (disunify t1 t2 st)
+         (#f          #f)
+         (`(,st . #f) (pause st (=/= t1 t2)))
```
Comment (gregr) [inline tools.rkt:81]: `disunify` should never return a pair.

**Review by gregr — CHANGES_REQUESTED** [review]
Looks good so far.

I'd prefer if you adjust your editor to use 2-space indents in situations where it's currently using 1-space, or else there will be whitespace-churning changes when other people edit this code in the future.

`fresh` should be indented like a binding form, as with `define`, `let`, `lambda`, etc.

Your indentation fixes for `if` are fine, though.

**Review by gregr — CHANGES_REQUESTED** [review]
Thanks, the indentation looks better now.

I know it's not code you were involved with, but I noticed this definition is inconsistent with how `walk*` is defined:
```
(define (walked-term t st) (walk* t (state-sub st)))
```

Either `walk*` should be changed to take just a substitution, which is all it really needs, or all callers of `walk*` need to pass it a state.  Although fixing `walk*` is nicer in the long run, I'm fine with the expedient approach if you prefer.

**Comment by pinnouse** [comment]
Updated `walk*` to take only substitutions, and all calls to `walk*` to use to pass the substitution instead of state.

**Review by gregr — APPROVED** [review]
Looks good.  Thanks for fixing `walk*` too.

**Pull request merged**